src/main/scala/mesosphere/marathon/state/Container.scala

```diff
@@ -1,143 +1,141 @@
       )
+
+    def unapply(volume: Volume): Option[(String, Option[String], Mesos.Volume.Mode, Option[PersistentVolumeInfo])] =
+      volume match {
+        case persistentVolume: PersistentVolume =>
+          Some((persistentVolume.containerPath, None, persistentVolume.mode, Some(persistentVolume.persistent)))
+        case dockerVolume: DockerVolume =>
+          Some((dockerVolume.containerPath, Some(dockerVolume.hostPath), dockerVolume.mode, None))
+      }
   }
 
   /**
-    * Docker-specific container parameters.
+    * A volume mapping either from host to container or vice versa.
+    * Both paths can either refer to a directory or a file.  Paths must be
+    * absolute.
     */
-  case class Docker(
-      image: String = "",
-      network: Option[mesos.ContainerInfo.DockerInfo.Network] = None,
-      portMappings: Option[Seq[Docker.PortMapping]] = None,
-      privileged: Boolean = false,
-      parameters: Seq[Parameter] = Nil,
-      forcePullImage: Boolean = false) {
-
-    def toProto(): Protos.ExtendedContainerInfo.DockerInfo = {
-      val builder = Protos.ExtendedContainerInfo.DockerInfo.newBuilder
-
-      builder.setImage(image)
-
-      network foreach builder.setNetwork
-
-      portMappings.foreach { pms =>
-        builder.addAllPortMappings(pms.map(_.toProto).asJava)
-      }
-
-      builder.setPrivileged(privileged)
-
-      builder.addAllParameters(parameters.map(_.toProto).asJava)
-
-      builder.setForcePullImage(forcePullImage)
-
-      builder.build
+  case class DockerVolume(
+    containerPath: String,
+    hostPath: String,
+    mode: Mesos.Volume.Mode)
+      extends Volume
+
+  object DockerVolume {
+
+    implicit val dockerVolumeValidator = validator[DockerVolume] { vol =>
+      vol.containerPath is notEmpty
+      vol.hostPath is notEmpty
+      vol.mode is oneOf(Mode.RW, Mode.RO)
     }
+  }
 
-    def toMesos(): mesos.ContainerInfo.DockerInfo = {
-      val builder = mesos.ContainerInfo.DockerInfo.newBuilder
-
-      builder.setImage(image)
-
-      network foreach builder.setNetwork
-
-      portMappings.foreach { pms =>
-        builder.addAllPortMappings(pms.map(_.toMesos).asJava)
-      }
-
-      builder.setPrivileged(privileged)
-
-      builder.addAllParameters(parameters.map(_.toProto).asJava)
-
-      builder.setForcePullImage(forcePullImage)
+  case class PersistentVolumeInfo(size: Long)
 
-      builder.build
+  object PersistentVolumeInfo {
+    implicit val persistentVolumeInfoValidator = validator[PersistentVolumeInfo] { info =>
+      (info.size > 0) is true
     }
+  }
 
+  case class PersistentVolume(
+    containerPath: String,
+    persistent: PersistentVolumeInfo,
+    mode: Mesos.Volume.Mode)
+      extends Volume
+
+  object PersistentVolume {
+    import org.apache.mesos.Protos.Volume.Mode
+    implicit val persistentVolumeValidator = validator[PersistentVolume] { vol =>
+      vol.containerPath is notEmpty
+      vol.persistent is valid
+      vol.mode is equalTo(Mode.RW)
```

Comment:
Mesos Persistent Volumes only allows `RW` – other volume types allow `RO` as well, so I guess it makes sense to enforce this here (we can't validate this constraint via the json schema). Plus, I want Alice to receive a validation error if she consciously tries to set `RO`.